ext/liquid_c/vm_assembler.c

```diff
@@ -116,3 +139,103 @@ void vm_assembler_add_push_literal(vm_assembler_t *code, VALUE literal)
         break;
     }
 }
+
+static void ensure_parsing(vm_assembler_t *code)
+{
+    if (!code->parsing)
+        rb_raise(rb_eRuntimeError, "cannot extend code after it has finished being compiled");
+}
+
+void vm_assembler_add_evaluate_expression_from_ruby(vm_assembler_t *code, VALUE code_obj, VALUE expression)
+{
+    ensure_parsing(code);
+
+    if (RB_SPECIAL_CONST_P(expression)) {
+        vm_assembler_add_push_literal(code, expression);
+        return;
+    }
+
+    switch (RB_BUILTIN_TYPE(expression)) {
+        case T_DATA:
+            if (RBASIC_CLASS(expression) == cLiquidCExpression) {
+                vm_assembler_concat(code, &((expression_t *)DATA_PTR(expression))->code);
+                vm_assembler_remove_leave(code);
+                return;
+            }
+            break;
+        case T_OBJECT:
+        {
+            VALUE klass = RBASIC_CLASS(expression);
+            if (klass == cLiquidVariableLookup || klass == cLiquidRangeLookup) {
+                rb_funcall(expression, id_compile_evaluate, 1, code_obj);
+                return;
+            }
+            break;
+        }
+        default:
+            break;
+    }
+    vm_assembler_add_push_const(code, expression);
+}
+
+void vm_assembler_add_find_variable_from_ruby(vm_assembler_t *code, VALUE code_obj, VALUE expression)
+{
+    ensure_parsing(code);
+
+    if (RB_TYPE_P(expression, T_STRING)) {
+        vm_assembler_add_find_static_variable(code, expression);
+    } else {
+        vm_assembler_add_evaluate_expression_from_ruby(code, code_obj, expression);
+        vm_assembler_add_find_variable(code);
+    }
+}
+
+void vm_assembler_add_lookup_command_from_ruby(vm_assembler_t *code, VALUE command)
+{
+    StringValue(command);
+    ensure_parsing(code);
+    vm_assembler_require_stack_args(code, 1);
+
+    vm_assembler_add_lookup_command(code, command);
+}
+
+void vm_assembler_add_lookup_key_from_ruby(vm_assembler_t *code, VALUE code_obj, VALUE expression)
+{
+    ensure_parsing(code);
+    vm_assembler_require_stack_args(code, 1);
+
+    if (RB_TYPE_P(expression, T_STRING)) {
+        vm_assembler_add_lookup_const_key(code, expression);
+    } else {
+        vm_assembler_add_evaluate_expression_from_ruby(code, code_obj, expression);
+        vm_assembler_add_lookup_key(code);
+    }
+}
+
+void vm_assembler_add_new_int_range_from_ruby(vm_assembler_t *code)
+{
+    ensure_parsing(code);
+    vm_assembler_require_stack_args(code, 2);
+    vm_assembler_add_new_int_range(code);
+}
+
+void vm_assembler_add_hash_new_from_ruby(vm_assembler_t *code, VALUE hash_size_obj)
+{
+    ensure_parsing(code);
+    unsigned int hash_size = NUM2USHORT(hash_size_obj);
+    if (hash_size > 255)
+        rb_enc_raise(utf8_encoding, cLiquidSyntaxError, "Hash literal has too many keys");
```

Comment:
Can we move this check to `vm_assembler_add_hash_new` just like `vm_assembler_add_filter`?

Reply:
Yes, although, we still need the check in `keyword_arg_count > 255` in `try_variable_strict_parse` so that we can provide a better error message.